### PR TITLE
Added Dynamic -X axis for empty response, Visualization for Constant Values , Fixed Step Size Error When Switching Between Time-Frames

### PR DESCRIPTION
**Pull request by Divyanshu11011**
# Description
Here's an improved version of your text:

Added dynamic X-axis for empty responses
Enhanced visualization for constant values
Fixed step size error when switching between time frames

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [x ] I have self-reviewed this PR.
- [x ] I have removed all print-debugging and commented-out code that should not be merged.
- [ x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x ] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

# Snapshot
image
image

### CODE REVIEW

File: static/js/metrics-explorer.js [hunk 1]
```diff
@@ -969,51 +977,29 @@ function initializeChart(canvas, seriesData, queryName, chartType) {
...
-    var existingContainer = $(`.metrics-graph[data-query="${queryName}"]`)
+    var existingContainer = $(`.metrics-graph[data-query="${queryName}"]`);
     if (existingContainer.length === 0){
         var visualizationContainer = $(`
         <div class="metrics-graph" data-query="${queryName}">
             <div class="query-string">${queryString}</div>
             <div class="graph-canvas"></div>
         </div>`);
 
-        // Determine where to insert the new container
```
Comment (sonamgupta21) [inline static/js/metrics-explorer.js:981]: Why are these codes removed? It broke some features on the metrics and alerts screen.

File: static/js/metrics-explorer.js [hunk 2]
```diff
@@ -969,51 +977,29 @@ function initializeChart(canvas, seriesData, queryName, chartType) {
...
         existingContainer.find('.query-string').text(queryString);
         var canvas = $('<canvas></canvas>');
         existingContainer.find('.graph-canvas').empty().append(canvas);
     }
 
     var lineChart = initializeChart(canvas, seriesData, queryName, chartType);
-
     lineCharts[queryName] = lineChart;
     updateGraphWidth();
-    mergeGraphs(chartType)
```
Comment (sonamgupta21) [inline static/js/metrics-explorer.js:1014]: Same here.

File: static/js/metrics-explorer.js [hunk 3]
```diff
@@ -1500,14 +1539,16 @@ async function getMetricsDataForFormula(formulaId, formulaDetails){
         rawTimeSeriesData = res;
     }
 
-    const chartData = await convertDataForChart(rawTimeSeriesData)
-    if(isAlertScreen){
```
Comment (sonamgupta21) [inline static/js/metrics-explorer.js:1504]: Same here.

**Review by sonamgupta21 — COMMENTED** [review]
x-axis isn't updating correctly when the time-picker changes
image

**Review by sonamgupta21 — COMMENTED** [review]
image

The x-axis is not displaying correctly. 
There are many changes due to formatting, making it very hard to review. It would be helpful if you could create a PR with the required changes only.

File: static/js/metrics-explorer.js [hunk 4]
```diff
@@ -978,96 +1008,173 @@ function prepareChartData(seriesData, chartDataCollection, queryName) {
...
+
+function generateEmptyTimeLabels(startDate) {
+    const endTime = new Date();
+    let startTime, stepSize, unit;
+
+    // Determine start time and step size based on selected time frame
+    if (startDate.includes("now-5m")) {
+        startTime = new Date(endTime.getTime() - 5 * 60 * 1000);
+        stepSize = 1 * 60 * 1000; // 1 minute
+        unit = 'minute';
```
[the remark this answers is not in the document]
Comment (Divyanshu11011) [reply static/js/metrics-explorer.js:1118]: The generated labels from generateEmptyTimeLabels are used in the initializeChart function to set up the chart's x-axis labels when there is no data available.

1) convertDataForChart processes actual data dynamically, while generateEmptyTimeLabels ensures a regular structure when there's no data.
2) generateEmptyTimeLabels provides time labels even when data is missing, which convertDataForChart can't do without data points.

[the remark this answers is not in the document]
Comment (sunitakawane) [reply static/js/metrics-explorer.js:1118]: I do not see unit variable beign used anywhere else. It is just assigned a value and not being used

[the remark this answers is not in the document]
Comment (sunitakawane) [reply static/js/metrics-explorer.js:1118]: For same timerange both graphs should have same labels

[the remark this answers is not in the document]
Comment (Divyanshu11011) [reply static/js/metrics-explorer.js:1118]: Yes - we can remove the unit-assignment part from the generateEmptyTimeLabels function , i will just push the change

[the remark this answers is not in the document]
Comment (Divyanshu11011) [reply static/js/metrics-explorer.js]: For the same time range, the x-axes are generated dynamically by the Chart.js library. Even if I select the same time frame for a metric, there may be differences in the x-axis. -  here is a snapshot of playground -
image

**Comment by sunitakawane** [comment]
Does not work with custom timerange 
Screenshot 2024-07-15 at 6 13 38 PM

**Comment by Divyanshu11011** [comment]
> Does not work with custom timerange Screenshot 2024-07-15 at 6 13 38 PM

Fixed it to work with - Custom time frames as well

**Review by sunitakawane — APPROVED** [review]

**Pull request merged**